developer/src/kmcmplib/tests/gtest-compiler-test.cpp

```diff
@@ -808,6 +1072,80 @@ TEST_F(CompilerTest, GetXStringImpl_type_i_test) {
     EXPECT_EQ(CERR_InvalidIndex, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
 }
 
+// tests strings starting with 'o'
+TEST_F(CompilerTest, GetXStringImpl_type_o_test) {
+    KMX_WCHAR tstr[128];
+    fileKeyboard.version = VERSION_80;
+    KMX_WCHAR str[LINESIZE];
+    KMX_WCHAR output[GLOBAL_BUFSIZE];
+    PKMX_WCHAR newp = nullptr;
+    PFILE_STORE file_store = new FILE_STORE[100];
+    fileKeyboard.cxStoreArray = 3u;
+    fileKeyboard.dpStoreArray = file_store;
+    file_store[1].fIsStore = TRUE;
+    u16cpy(file_store[0].szName, u"a");
+    u16cpy(file_store[1].szName, u"b");
+    u16cpy(file_store[2].szName, u"c");
+
+    // CERR_InvalidToken
+    u16cpy(str, u"opq");
+    EXPECT_EQ(CERR_InvalidToken, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, CERR_OutsInVirtualKeySection *** TODO ***
+
+    // outs, no close delimiter => NULL
+    u16cpy(str, u"outs(");
+    EXPECT_EQ(CERR_InvalidOuts, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, empty delimiters => empty string
+    u16cpy(str, u"outs()");
+    EXPECT_EQ(CERR_InvalidOuts, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, space in delimiters (see I11814, I11937, #11910, #11894, #11938)
+    u16cpy(str, u"outs( )");
+    EXPECT_EQ(CERR_InvalidOuts, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, CERR_StoreDoesNotExist
+    u16cpy(str, u"outs(d)");
+    EXPECT_EQ(CERR_StoreDoesNotExist, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, CERR_StoreDoesNotExist, space before store
+    u16cpy(str, u"outs( d)");
+    EXPECT_EQ(CERR_StoreDoesNotExist, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, CERR_StoreDoesNotExist, space after store
+    u16cpy(str, u"outs(d )");
+    EXPECT_EQ(CERR_StoreDoesNotExist, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // outs, CERR_OutsTooLong
+    PKMX_WCHAR dpString = (PKMX_WCHAR)u"abc";
+    file_store[1].dpString = dpString; // length 4 => max should be > 4, otherwise a CERR_OutsTooLong is emitted
+    int max = u16len(dpString) + 1; // 4, including terminating '\0'
+    u16cpy(str, u"outs(b)");
+    EXPECT_EQ(CERR_OutsTooLong, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, max, 0, &newp, FALSE)); // max reduced to force error
```

Comment:
So, I realize that the code being tested isn't yours... but man, that is **_not_** an intuitive pattern for `max`.  In the worst case, I'd expect max=4 to be perfectly fine with 3 chars + 1 null terminator.  4 is within a max of 4, after all.  (I can understand not making the null terminator implicit due to how C / C++ strings work.)

Reply:
Sure. Looks like an off-by-one in calculation, but it's off-by-one in a conservative way so not really any harm done. If you are looking for intuitive code, I suggest you don't look at the compiler.

Reply:
Yeah, so many problems with buffer management. Yay C++. My high-level fix is an attempt to mitigate rather than resolve, as we slowly inch towards a rewrite one day.